### PR TITLE
CRM-17633: Changes to support WP in it's own directory. 

### PR DESCRIPTION
**Pull request by kcristiano**
Overview
----------------------------------------
Allow for civicrn.settings.extra.php for WP. Create template for civicrm.settings.extra.php. Update install/civicrm.php for all needed params for different common install scenarios. Fix linting issues in civicrm.php.

CRM 16421 CRM 17633 - update CRM_Utils_System_WordPress to allow for common install configurations

Replaces https://github.com/civicrm/civicrm-core/pull/10214/
Before
----------------------------------------
WP install will fail if not installed in tyhe webroot.  This supports https://wiki.civicrm.org/confluence/display/CRM/WordPress+installed+in+its+own+directory+issues

https://issues.civicrm.org/jira/browse/CRM-17633

After
----------------------------------------
Create civicrm.setting.phpo file based on WP directory structure in use

Comments
----------------------------------------
This is a rebase of the prior PR.   Mertges 2 weeks ago had broken the original

Depends on https://github.com/civicrm/civicrm-wordpress/pull/105

### CODE REVIEW

**Comment by kcristiano** [comment]
jenkins test this please

**Comment by eileenmcnaughton** [comment]
Jenkins:

  [GitScan\Exception\ProcessErrorException]                                           
  Process failed:                                                                     
  [[ COMMAND: git apply --check --ignore-whitespace ]]                                
  [[ CWD: /home/jenkins/buildkit/build/core-11031-813e8/sites/all/modules/civicrm ]]  
  [[ EXIT CODE: 1 ]]                                                                  
  [[ STDOUT ]]                                                                        
  [[ STDERR ]]                                                                        
  error: patch failed: templates/CRM/common/civicrm.settings.php.template:28          
  error: templates/CRM/common/civicrm.settings.php.template: patch does not apply

**Comment by kcristiano** [comment]
@civicrm-builder retest this please

**Comment by kcristiano** [comment]
@civicrm-builder retest this please

**Comment by alifrumin** [comment]
I tested this pr and https://github.com/civicrm/civicrm-wordpress/pull/105 on a wordpress install with wordpress in the root site directory and with wordpress in its own directory and did not get any issues regarding directories or resource urls.

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by kcristiano** [comment]
@civicrm-builder test this please

**Comment by kcristiano** [comment]
@eileenmcnaughton I am getting:

`
CRM_Utils_GeocodeTest::testStateProvinceFormat
Failed asserting that false is true.

/home/jenkins/buildkit/build/core-11031-85il8/sites/all/modules/civicrm/tests/phpunit/CRM/Utils/GeocodeTest.php:19
/home/jenkins/buildkit/build/core-11031-85il8/sites/all/modules/civicrm/tests/phpunit/CiviTest/CiviUnitTestCase.php:184
/home/jenkins/buildkit/bin/phpunit4:545
`
Do you know if there are issues with Jenkins?  I can't imagine how these changes would trigger a failure there.  These are all specific to the installer and only in the WP context.  I was hoping to finally get these in for 4.7.26-rc testing as the best testing for the installer is via tarballs.

**Comment by MegaphoneJon** [comment]
@kcristiano It's most likely the issue discussed yesterday here: https://chat.civicrm.org/civicrm/pl/51tgxt8zajy3mjrcub69d5nnze

**Comment by kcristiano** [comment]
retest this please

**Comment by eileenmcnaughton** [comment]
@kcristiano tests are passing now & @alifrumin has tested. Code looks good. 

I feel comfortable that there is no impact outside WP dirs - giving this merge-ready in the hope @christianwach will take a look & confirm he is good with this but not sure it needs to be blocked on that given the input already given

**Comment by kcristiano** [comment]
Thanks @eileenmcnaughton Can we also mark https://github.com/civicrm/civicrm-wordpress/pull/105 this one as well?  Thye are related and my hope is to get these in the next RC.   As they affect the installer, the RC is a perfect way to validate my testing.

**Comment by eileenmcnaughton** [comment]
hmm - we don't have those labels on that repo

File: Civi/Core/Paths.php [hunk 1]
```diff
@@ -55,6 +56,24 @@ public function __construct() {
...
+      })
+      ->register('wp.frontend', function () use ($paths) {
+        $config = \CRM_Core_Config::singleton();
+        $suffix = defined('CIVICRM_UF_WP_BASEPAGE') ? CIVICRM_UF_WP_BASEPAGE : $config->wpBasePage;
+        return array(
+          'url' => $paths->getVariable('wp.frontend.base', 'url') . $suffix,
+        );
+      })
+      ->register('wp.backend.base', function () {
+        return array('url' => CIVICRM_UF_BASEURL . 'wp-admin/');
```
Comment (christianwach) [inline Civi/Core/Paths.php]: @kcristiano Are you sure that `wp-admin/` should be hard-coded here? The [`admin_url`](https://developer.wordpress.org/reference/hooks/admin_url/) filter allows this to be overridden, so if this path has to be correct without bootstrapping WordPress, then perhaps there needs to be a setting for it somewhere? OTOH, perhaps I'm raising an issue that's outside the scope of this PR.

Comment (kcristiano) [reply Civi/Core/Paths.php]: @christianwach I see the point and had originally done that.  In one of the iterations (this PR in various forms dates back to Ft Collins Sprint in June 2016), @totten  made the followimng PRs that have been added:  https://github.com/kcristiano/civicrm-core/commit/48c4fe6f4a9c61f1bc701a2f81377adab7f3c4df  and https://github.com/kcristiano/civicrm-core/commit/8b0c250621cf2a433ba2c6c4d121715671b8f024    This section https://github.com/kcristiano/civicrm-core/commit/8b0c250621cf2a433ba2c6c4d121715671b8f024#diff-46f9229e858ccffc107ef60a91809c43L286  removed the call to admin_url().   I attempted to add back, but had errors depending on the configuration.

As I would much rather have this in and then improve than hold it back even longer, I consider switching to admin_url() as a TODO.   We can create an issue for that and fix it after these changes are tested through an RC.   As this is an installer patch and it affects both civicrm-core and civicrm-wordpress we'll get more testing once it hits RC.

Comment (christianwach) [reply Civi/Core/Paths.php]: @kcristiano Gotcha. Seems like a good idea.

**Comment by eileenmcnaughton** [comment]
Ok - I think gotcha translates to 'yes this is ready to merge IMHO' - merging this & the packages one to hit the rc.

**Pull request merged**